### PR TITLE
Publicize necessary functionality for creating diversifiers and notes from data

### PR DESCRIPTION
**Pull request by AloeareV**
One use-case is to enable wallets to store notes and diversifiers on disk.

### CODE REVIEW

**Review by therealyingtong — APPROVED** [review]
utACK

**Comment by zancas** [comment]
Pinging on this PR...   looks like it's almost ready?

**Comment by zancas** [comment]
Hiya @therealyingtong it looks like merging this PR will also unblock this development effort:

https://github.com/zcash/orchard/pull/346

Is it the case that it simply lacks a second review?

**Comment by AloeareV** [comment]
Made some changes, in line with daira's feedback on #346

File: src/note.rs
Comment (daira) [inline src/note.rs:39]:
```suggestion
    /// Read a note's random seed from bytes, given the note's nullifier.
    /// Returns `None` if the nullifier is not for the same note as the seed.
```

File: src/note.rs
Comment (daira) [inline src/note.rs:115]:
```suggestion
    /// Create a Note from its component parts. Returns `None` if a valid
    /// [`NoteCommitment`] cannot be derived from the note.
```

File: src/keys.rs
Comment (daira) [inline src/keys.rs:555]:
```suggestion
    /// Read a diversifier from a byte array.
```

**Review by daira — APPROVED** [review]
utACK modulo @str4d's suggestions for the changelog.

**Review by str4d — CHANGES_REQUESTED** [review]
We discussed these API changes, and decided that this is consistent with how we want the public API to be expanded.

File: src/note.rs
Comment (AloeareV) [reply src/note.rs]: Thanks! Comments and changelog updated!

**Comment by str4d** [comment]
CI is failing because the new commit that was pushed triggered a test of this merged into current `main`, which is currently broken. Once #358 is merged, we will re-run CI.

**Comment by nuttycom** [comment]
> CI is failing because the new commit that was pushed triggered a test of this merged into current `main`, which is currently broken. Once #358 is merged, we will re-run CI.

Since what is tested is the branch state, not the merged state, this PR will need to be rebased again, right?

**Comment by str4d** [comment]
> This PR will need to be rebased again, right?

I think it just needs to be pushed again; the PR contents doesn't have to change. I have squashed the fixup commits into the original 3 locally to clean up the history in a way that doesn't alter this PR's contents, but because Zingo Labs have disabled the ability for us to push to their branches (usually this is a setting that is default-enabled when creating a PR, but maybe there's some Zingo Labs org setting that is inverting the permission), I can't update this PR.

**Comment by AloeareV** [comment]
> but because Zingo Labs have disabled the ability for us to push to their branches (usually this is a setting that is default-enabled when creating a PR, but maybe there's some Zingo Labs org setting that is inverting the permission), I can't update this PR.

https://github.com/community/community/discussions/5634

It seems this feature doesn't exist for organization-owned repos, only user-owned ones. I'll do a similar squash.

**Comment by str4d** [comment]
> It seems this feature doesn't exist for organization-owned repos

Aha, that explains it! I had assumed it was because org-owned repos had extra config options to lock down permissions, not that they had fewer config options.

File: src/note.rs
Comment (str4d) [inline src/note.rs:133]: It occurs to me now that a potentially simpler API for downstream users would be for the public `from_parts` to take `rseed` as a `[u8; 32]` and not expose the `RandomSeed` type externally at all (currently the only thing you can do with it publicly is wrap it around a byte array, and I don't see this changing in future). The `RandomSeed::from_bytes` check would then be done internally to `Note::from_parts` as:
```rust
impl Note {
    pub fn from_parts(
        recipient: Address,
        value: NoteValue,
        rho: Nullifier,
        rseed: [u8; 32],
    ) -> CtOption<Self> {
        RandomSeed::from_bytes(rseed, &rho)
            .and_then(|rseed| Self::from_parts_internal(recipient, value, rho, rseed))
    }

    fn from_parts_internal(
        recipient: Address,
        value: NoteValue,
        rho: Nullifier,
        rseed: RandomSeed,
    ) -> CtOption<Self> {
        ...
    }

    pub fn rseed(&self) -> &[u8; 32] {
        self.rseed.as_bytes()
    }
}
```

On the other hand, having the `RandomSeed` type exist may make it easier for crate users to keep it typed in the space between where they interact with their database and where they interact with `Note`. Then again, there is nothing stopping them only using `RandomSeed` immediately before `Note::from_parts`, which makes the type safety completely pointless in their usage:
```rust
let note = RandomSeed::from_bytes(rseed, &rho)
    .and_then(|rseed| Note::from_parts(recipient, value, rho, rseed))
````

**Review by str4d — APPROVED** [review]
utACK 0800d23fe729ebcf1ea129688bfe67f9d2d7f29d

File: src/note.rs
Comment (zancas) [reply src/note.rs:133]: Can we call the local variable name on the line that starts with `.and_then` something other than `rseed`?

Comment (nuttycom) [reply src/note.rs:133]: I think that this is okay as it is currently; it makes the role of `rho` clear in constructing the `RandomSeed` value.

**Review by nuttycom — APPROVED** [review]
utACK https://github.com/zcash/orchard/commit/0800d23fe729ebcf1ea129688bfe67f9d2d7f29d

File: src/note.rs
Comment (AloeareV) [reply src/note.rs:133]: The `RandomSeed` being constructed beforehand makes it more obvious to the consumer whether the note construction went wrong due to a mismatched rho/rseed, or something else. I don't know how to weigh that against the ergonomics of it.

**Pull request merged**
